src/main.rs

```diff
@@ -559,20 +559,22 @@ extern "C" fn primary_init(hart_id: u64, fdt_addr: u64) -> CpuLaunch {
 
     HOST_VM.call_once(|| host);
 
-    // For now, reset the stack at the top.
-    // Safe because we trust the linker placed these symbols correctly.
-    let sp = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
-
     // Return launch parameters for this CPU.
+    let this_cpu = PerCpu::this_cpu();
     CpuLaunch {
-        satp: PerCpu::this_cpu().page_table().satp(),
-        sp,
+        satp: this_cpu.page_table().satp(),
+        sp: this_cpu.stack_top(),
     }
 }
 
 /// Start salus on the bootstrap CPU. This is called once we switch to the hypervisor page-table.
 #[no_mangle]
 extern "C" fn primary_main() {
+    let sp: u64;
+    // Safe since there is no memory access.
+    unsafe { core::arch::asm!("mv {rs}, sp", rs = out(reg) sp) };
```

Comment:
the asm shuld have "nomem" to indicate the absence of a memory clobber then. (here and below)

Reply:
Sure.